### PR TITLE
Bump to xamarin/Java.Interop/main@bbaeda6f

### PR DESCRIPTION
**Pull request by jonpryor**
Changes: https://github.com/xamarin/java.interop/compare/9e0a4690adb71c04cd88a5b54bea3c3f8da73cc0...bbaeda6f698369bdd48bfc2dd1a32a41c9d23229
    
  * xamarin/java.interop@bbaeda6f: [Java.Interop] Support Desugar + interface static methods (xamarin/java.interop#1077)

Context: https://github.com/xamarin/java.interop/commit/bbaeda6f698369bdd48bfc2dd1a32a41c9d23229
Context: https://github.com/xamarin/java.interop/commit/1f27ab552d03aeb74cdc6f8985fcffbfdb9a7ddf
Context: f6f11a5a797cd8602854942dccb0f2b1db57c473

[Desugaring][0] is the process of rewriting Java bytecode so that Java 8+ constructs can be used on Android pre-7.0 (API-24), as API-24 is the Android version which added native support for Java 8 features such as [interface default methods][1].

One of the implications of desugaring is that methods can "move"; consider this Java interface:

	package example;
	public interface StaticMethodsInterface {
	    static int getValue() { return 3; }
	}

Java.Interop bindings will attempt to invoke the `getValue().I` method on the type `example/StaticMethodsInterface`:

	public partial interface IStaticMethodsInterface : IJavaObject, IJavaPeerable {
	    private static readonly JniPeerMembers _members = new XAPeerMembers ("example/StaticMethodsInterface", typeof (IStaticMethodsInterface), isInterface: true);
	    static unsafe int Value {
	        get {
	            const string __id = "getValue.()I";
	            var __rm = _members.StaticMethods.InvokeInt32Method (__id, null);
	            return __rm;
	        }
	    }
	}

The problem is that, after Desugaring, the Java side *actually* looks like this:

	package example;
	public interface StaticMethodsInterface {
	}
	public class StaticMethodsInterface$-CC {
	    public static int getValue() {return 3;}
	}

Commits xamarin/java.interop@1f27ab55 and f6f11a5a added partial runtime support for this scenario via
`AndroidTypeManager.GetStaticMethodFallbackTypesCore()`, which would attempt to lookup types with a `$-CC` suffix.

While this was a good start, it wasn't ever actually *tested* end-to-end.  Consequently, instead of *working*, this would instead cause the process to *abort*:

	JNI DETECTED ERROR IN APPLICATION: can't call static int example.StaticMethodsInterface$-CC.getValue() with class java.lang.Class<example.StaticMethodsInterface>
	    in call to CallStaticIntMethodA
	    from void crc….MainActivity.n_onCreate(android.os.Bundle)

Oops.

xamarin/java.interop#1077 improves our runtime support for invoking *`static`* methods on interfaces.

Add a new `XASdkDeployTests.SupportDesugaringStaticInterfaceMethods()` test which performs an on-device, end-to-end invocation of a static method on a Java interface, to ensure that things *actually* work.

*Note*: if `$(SupportedOSPlatformVersion)` is 24 or higher, this test will work even without xamarin/java.interop#1077, as Desugaring is *disabled* in that case.

[0]: https://developer.android.com/studio/write/java8-support#library-desugaring
[1]: https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html

### CODE REVIEW

File: tests/MSBuildDeviceIntegration/Tests/XASdkDeployTests.cs [hunk 1]
```diff
@@ -155,6 +155,59 @@ public void TypeAndMemberRemapping ([Values (false, true)] bool isRelease)
...
+							{ "Bind", "True" },
+						},
+					},
+				},
+			};
+
+			// Note: To properly test, Desugaring must be *enabled*, which requires that
+			// `$(SupportedOSPlatformVersion)` be *less than* 23.  21 is currently the default,
+			// but set this explicitly anyway just so that this implicit requirement is explicit.
+			proj.SetProperty (proj.ReleaseProperties, "SupportedOSPlatformVersion", "21");
```
Comment (dellis1972) [inline tests/MSBuildDeviceIntegration/Tests/XASdkDeployTests.cs:182]: Do these tests still run under legacy? If so will this work?

Comment (dellis1972) [reply tests/MSBuildDeviceIntegration/Tests/XASdkDeployTests.cs:182]: Never mind these are the Sdk tests.

**Review by dellis1972 — APPROVED** [review]

**Comment by jonpryor** [comment]
"interestingly", the new *xamarin-android* [SupportDesugaringStaticInterfaceMethods](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7349165&view=ms.vss-test-web.build-test-results-tab&runId=68887792&resultId=100022&paneView=debug) test passes, while the *Java.Interop* test [`DesugarInterfaceStaticMethod()`  test *fails*](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7349165&view=ms.vss-test-web.build-test-results-tab&runId=68885339&resultId=100494&paneView=debug):

```
Java.Lang.NoSuchMethodError : no static method "Lcom/xamarin/interop/AndroidInterface;.getClassName()Ljava/lang/String;"
   at Java.Interop.JniEnvironment.StaticMethods.GetStaticMethodID(JniObjectReference , String , String )
   at Java.Interop.JniType.GetStaticMethod(String , String )
   at Java.Interop.JniPeerMembers.JniStaticMethods.GetMethodInfo(String , String )
   at Java.Interop.JniPeerMembers.JniStaticMethods.GetMethodInfo(String )
   at Java.Interop.JniPeerMembers.JniStaticMethods.InvokeObjectMethod(String , JniArgumentValue* )
   at Java.InteropTests.IAndroidInterface.getClassName()
   at Java.InteropTests.JniPeerMembersTests.DesugarInterfaceStaticMethod()
   at System.Reflection.MethodInvoker.InterpretedInvoke(Object , IntPtr* )
   at System.Reflection.MethodInvoker.Invoke(Object , IntPtr* , BindingFlags )
  --- End of managed Java.Lang.NoSuchMethodError stack trace ---
java.lang.NoSuchMethodError: no static method "Lcom/xamarin/interop/AndroidInterface;.getClassName()Ljava/lang/String;"
	at crc643df67da7b13bb6b1.TestInstrumentation_1.n_onStart(Native Method)
	at crc643df67da7b13bb6b1.TestInstrumentation_1.onStart(TestInstrumentation_1.java:27)
	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
```

[`$(SupportedOSPlatformVersion)` for `Mono.Android.NET-Tests.csproj`](https://github.com/xamarin/xamarin-android/blob/37a1758dd622a8ffc8aa1a53bedb7f5cfbe78a0d/tests/Mono.Android-Tests/Runtime-Microsoft.Android.Sdk/Mono.Android.NET-Tests.csproj#L8) is 21, which should enable *real* Desugar, so… what's wrong?

**Comment by jonpryor** [comment]
[following up](https://github.com/xamarin/xamarin-android/pull/7799#issuecomment-1439008221)…

If I look at `base/dex/classes.dex` within [`Mono.Android.NET_Tests-Signed.aab`](https://artprodscussu1.artifacts.visualstudio.com/A011b8bdf-6d56-4f87-be0d-0092136884d9/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2RldmRpdi9wcm9qZWN0SWQvMGJkYmM1OTAtYTA2Mi00YzNmLWIwZjYtOTM4M2Y2Nzg2NWVlL2J1aWxkSWQvNzM0OTE2NS9hcnRpZmFjdE5hbWUvVGVzdCtSZXN1bHRzKy0rQVBLcysuTkVUK1JlbGVhc2UrLSttYWNPUysoQXR0ZW1wdCsyKQ2/content?format=file&subPath=%2Fnet8.0-Release%2FMono.Android.NET_Tests-Signed.aab), I see:

```zsh
% dexdump base/dex/classes.dex| grep 'Class desc.*AndroidInter'
  Class descriptor  : 'Lcom/xamarin/interop/AndroidInterface;'
  Class descriptor  : 'Lcom/xamarin/interop/DesugarAndroidInterface$_CC;'
```

i.e. [`DesugarAndroidInterface$_CC`](https://github.com/xamarin/java.interop/pull/1077/files#diff-cd61a6dff62293bb846dbbbfbfead727a84a7c9477b332bccda4733dcfe1f7beR3) is included, but not `DesugarAndroidInterface$-CC` (note: `_` vs. `-`).

This is because [`AndroidInterface`](https://github.com/xamarin/java.interop/pull/1077/files#diff-168347fa2ab8e7e727291d7176b68d3fa9a2d08f343d57bd377a9a2fcbbf5ce6R6) is *wrong* for "Android with Desugar"; we need to instead have `com.xamarin.android.AndroidInterface` contain a static method so that Desugar has something to do.

It's thus somewhat disturbing that [SupportDesugaringStaticInterfaceMethods](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7349165&view=ms.vss-test-web.build-test-results-tab&runId=68887792&resultId=100022&paneView=debug) passes at all!

**Comment by jonpryor** [comment]
Grand unified commit message:

```markdown
Changes: https://github.com/xamarin/java.interop/compare/9e0a4690adb71c04cd88a5b54bea3c3f8da73cc0...bbaeda6f698369bdd48bfc2dd1a32a41c9d23229
    
  * xamarin/java.interop@bbaeda6f: [Java.Interop] Support Desugar + interface static methods (xamarin/java.interop#1077)

Context: https://github.com/xamarin/java.interop/commit/bbaeda6f698369bdd48bfc2dd1a32a41c9d23229
Context: https://github.com/xamarin/java.interop/commit/1f27ab552d03aeb74cdc6f8985fcffbfdb9a7ddf
Context: f6f11a5a797cd8602854942dccb0f2b1db57c473

[Desugaring][0] is the process of rewriting Java bytecode so that
Java 8+ constructs can be used on Android pre-7.0 (API-24), as
API-24 is the Android version which added native support for Java 8
features such as [interface default methods][1].

One of the implications of desugaring is that methods can "move";
consider this Java interface:

	package example;
	public interface StaticMethodsInterface {
	    static int getValue() { return 3; }
	}

Java.Interop bindings will attempt to invoke the `getValue().I`
method on the type `example/StaticMethodsInterface`:

	public partial interface IStaticMethodsInterface : IJavaObject, IJavaPeerable {
	    private static readonly JniPeerMembers _members = new XAPeerMembers ("example/StaticMethodsInterface", typeof (IStaticMethodsInterface), isInterface: true);
	    static unsafe int Value {
	        get {
	            const string __id = "getValue.()I";
	            var __rm = _members.StaticMethods.InvokeInt32Method (__id, null);
	            return __rm;
	        }
	    }
	}

The problem is that, after Desugaring, the Java side *actually* looks
like this:

	package example;
	public interface StaticMethodsInterface {
	}
	public class StaticMethodsInterface$-CC {
	    public static int getValue() {return 3;}
	}

Commits xamarin/java.interop@1f27ab55 and f6f11a5a added partial
runtime support for this scenario via
`AndroidTypeManager.GetStaticMethodFallbackTypesCore()`, which would
attempt to lookup types with a `$-CC` suffix.

While this was a good start, it wasn't ever actually *tested*
end-to-end.  Consequently, instead of *working*, this would instead
cause the process to *abort*:

	JNI DETECTED ERROR IN APPLICATION: can't call static int example.StaticMethodsInterface$-CC.getValue() with class java.lang.Class<example.StaticMethodsInterface>
	    in call to CallStaticIntMethodA
	    from void crc….MainActivity.n_onCreate(android.os.Bundle)

Oops.

xamarin/java.interop@bbaeda6f improves our runtime support for
invoking *`static`* methods on interfaces.

Add a new `XASdkDeployTests.SupportDesugaringStaticInterfaceMethods()`
test which performs an on-device, end-to-end invocation of a static
method on a Java interface, to ensure that things *actually* work.

*Note*: if `$(SupportedOSPlatformVersion)` is 24 or higher, this test
will work even without xamarin/java.interop#1077, as Desugaring is
*disabled* in that case.

The test `JniPeerMembersTests.DesugarInterfaceStaticMethod()` added
in xamarin/java.interop@bbaeda6f attempts to "fake" a Desugar
environment for testing on Desktop Java, and this "fakery" doesn't
work in the Android environment.  Fix execution on Android by updating
`AndroidRuntime.GetStaticMethodFallbackTypesCore()` to support type
remapping (f6f11a5a) -- which was overlooked/not considered --
such that the types returned are *after* calling
`AndroidRuntime.GetReplacementTypeCore()`, which looks up
`<replace-type/>` values.  This allows us to remap
`AndroidInterface` to `DesugarAndroidInterface$_CC`, allowing the
`DesugarInterfaceStaticMethod()` test to pass.

Update the `BuildTestJarFile` target so that it properly builds files
that contain `$` on Unixy platforms.

[0]: https://developer.android.com/studio/write/java8-support#library-desugaring
[1]: https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html
```

**Pull request merged**